File: languages/python/src/slop_guard/rules/smells/no_long_parameter_list.py
"""Reject functions that take more arguments than a caller can keep straight."""

from __future__ import annotations

from ...rule import FunctionNode, Rule

DEFAULT_MAX_PARAMETERS = 4
RECEIVERS = frozenset({"self", "cls", "mcs"})


class NoLongParameterList(Rule):
    name = "no-long-parameter-list"
    group = "smells"
    reference = "https://refactoring.guru/smells/long-parameter-list"
    description = "Disallow more than `max_parameters` parameters, 4 by default, excluding `self`."

    def visit_FunctionDef(self, node: FunctionNode) -> None:
        limit = self.threshold("max_parameters", DEFAULT_MAX_PARAMETERS)
        names = [parameter.arg for parameter in self.parameters(node) if parameter.arg not in RECEIVERS]
        if len(names) > limit:
            self.report(
                node,
                f"`{node.name}` takes {len(names)} parameters, past the {limit} this project "
                f"allows. Callers have to get every position right, and the list grows with every "
                f"new case. Pass the group that travels together as one object.",
            )
        self.generic_visit(node)

    visit_AsyncFunctionDef = visit_FunctionDef

File: languages/typescript/src/smells/rules/no-long-parameter-list.ts
import { defineRule } from "@oxlint/plugins";
import type { ESTree } from "@oxlint/plugins";

import { numberOption, thresholdSchema } from "../shared/options.ts";

const DEFAULT_MAX_PARAMETERS = 4;

type ParameterOwner =
	| ESTree.ArrowFunctionExpression
	| ESTree.Function
	| ESTree.TSMethodSignature;

/** Ban parameter lists long enough that callers must count positions. */
export const noLongParameterListRule = defineRule({
	meta: {
		type: "problem",
		docs: {
			description:
				"Disallow more than `maxParameters` parameters; values that travel together belong in one object.",
		},
		messages: {
			longParameterList:
				"This function takes {{count}} parameters, past the {{limit}} this project allows. Callers have to get every position right, and the list grows with every new case. Pass the group that travels together as one object.",
		},
		...thresholdSchema({ maxParameters: DEFAULT_MAX_PARAMETERS }),
	},
	createOnce(context) {
		const check = (node: ParameterOwner) => {
			const limit = numberOption(context.options, "maxParameters", DEFAULT_MAX_PARAMETERS);
			if (node.params.length <= limit) return;
			context.report({
				node,
				messageId: "longParameterList",
				data: { count: String(node.params.length), limit: String(limit) },
			});
		};

		return {
			ArrowFunctionExpression: check,
			FunctionDeclaration: check,
			FunctionExpression: check,
			TSDeclareFunction: check,
			TSMethodSignature: check,
		};
	},
});
